### ISSUE TITLE
Versioned release?

### ISSUE THREAD

**Issue by mr-c** [simd-everywhere/simde#50]
Hello, I'm considering packaging simde for Debian. Could you make a tagged release?

Thanks!

**Comment by nemequ** [comment]
SIMDe is really intended to be dropped directly into other projects (I would use a submodule or subtree, but just copying the simde/ directory should work fine, too), not installed system-wide, so I hadn't really planned on doing a versioned release. 

Now that I see what you're doing with the Debian packages, though, I think installing system-wide makes more sense than I had originally thought.  I'm guessing you don't want to carry a copy of SIMDe in the Debian packaging for each package, and not everyone is going to want to upstream the patches (though that would certainly be nice).

So, since you filed this issue I've been thinking about what to do about this.  I think I've more or less settled on a few points:

1. For now, I only want to install the x86/x86_64 headers.  I'm not ready to commit to any API stability for the others **yet**.
1. Before committing to a stable API I need to think about some of the parameters a bit more.  For example, there are a lot of "imm8" parameters in the x86/x86_64 APIs which use int, but it may be more appropriate to use `int_fast8_t`.  Unfortunately that's actually 8-bit, which isn't actually that fast, on my machine, but since they're immediate mode and everything should be inlined it shouldn't matter.
1. I'm not sure I'm ready to commit to the stability of union members other than `{i,u}{8,16,32,64}`, `float{32,64}`, and *maybe* `{u,i}32f` quite yet.
1. The current CMake file is really only designed to build the tests.  I need to move it to the top level and fix it up for other uses, such as installing the headers, tests, pkg-config, etc.
1, I"ll probably want to add SIMDE_AVAILABLE_SINCE_* macros to help people who want to maintain compatibility with specific versions of SIMDe, but that doesn't need to happen until after the first release.

Basically, it's going to take a bit of time, but this is on my radar.

**Comment by junaruga** [comment]
I'm considering packaging simde for Fedora project.

**Comment by nemequ** [comment]
Okay, I guess people are interested in an installed version.  I didn't expect that.

Once I finish converting to Meson I'll take a closer look at this.

**Comment by junaruga** [comment]
Thanks!

**Comment by nemequ** [comment]
I think it's time to start seriously considering this.  It's hard to get to the union members now; you have to go through `simde_*_to_private` to convert to a different type, which makes it pretty clear that you're using the private API when you do it.

Really the big thing is to make a decision on converting the rest of the types to standard types.  If anyone has any opinions on that I'd be interested.

**Comment by mr-c** [comment]
Sorry that I have no feedback on converting the rest of the types to standard types. Any further thoughts on a versioned release?

**Comment by junaruga** [comment]
@nemequ For example, when we release simde with the version 0.Y.Z, then when after we will complete the union members task, we lease simde as version 1.0.0, is there a problem for you?

At least in the case of [bowtie2](https://github.com/BenLangmead/bowtie2) using simde, it works well.

**Comment by nemequ** [comment]
I'm not concerned about the union anymore because for example the `__m128i` union is now called `simde__m128i_private`, and the `simde__m128i` we use as parameters/return values is usually typedef'd to either `__m128i` or `int_fast32_t __attribute__((__vector_size__(16)))`, so it's pretty clear that the union members aren't a stable API.  Even just to access them you are supposed to use a `simde__m128i_to_private(simde__m128i vec)` function (technically you can also do something like `reinterpret_cast<simde__m128i_private*>(&vec)`).

But yes, I think I'll start with a 0.5 or something, so if we really need to break the API we can.

Really I just need to make a decision about the types, and refactor the build system a bit.

I'll try to work on this soon; I'm pretty busy next week, so probably the week of the 9th.

**Comment by junaruga** [comment]
Sure. Thanks for explaining about the union issue.

> I'll try to work on this soon; I'm pretty busy next week, so probably the week of the 9th.

All right! Thanks for considering this.

**Comment by mr-c** [comment]
@nemequ Can we get a release prior to the start of GSoC? Then we will have some practice making releases and it will be easier to make a release right after GSoC is done, which will be nice for the mentees and good for everyone.

**Comment by junaruga** [comment]
@mr-c considering simde's Debian deb packaging, I would share the situation in Fedora RPM packaing. In case of Fedora's RPM packaging, maybe I will go a head even when simde is not versioned released.

Here is the command to download the simde's archive file by a given commit hash.

```
$ wget https://github.com/nemequ/simde/archive/53ed717a56ba6d136d1360a0b056ddf51825d5e3.tar.gz
```

Here is [the simde.spec file](https://github.com/junaruga/fedora-simde/blob/master/simde.spec) (RPM's recipe file) that I am preparing. You can see the simde's unit tests are executed with both system gcc and clang in the `%check` section.

> But yes, I think I'll start with a 0.5 or something, so if we really need to break the API we can.

I used `0.0.1` expecting it is less than the actually released version number.
@nemequ I expect "you will start with a 0.5 or something". :)

Here is the way to set the version and the release number by a upstream project's commit hash.

```
Version: 0.0.1
Release: 1.git%{short_commit_simde}%{?dist}
```

This way to set the release number by a commit hash is used in such as [moby-engine.spec](https://src.fedoraproject.org/rpms/moby-engine/blob/master/f/moby-engine.spec) and [podman.spec](https://src.fedoraproject.org/rpms/podman/blob/master/f/podman.spec).

**Comment by nemequ** [comment]
Do you need a build system to install the headers?

That's really the only blocker left on my list at this point.  I played with switching the immediate mode arguments to `int_fast*_t`, and hated it.  It was fine when the allowed ranges matched up with the types, bust most of the time an `int_fast8_t` was still too big and it just ended up looking awkward.

Other than that, I just want to get PPC64LE working on clang.  I have a patch for µnit working for that, I just need to do some additional testing on the µnit side to make sure it doesn't break anything.  I can probably be done with that tomorrow (unless it ends up horribly breaking something, but I think it's pretty safe).

> You can see the simde's unit tests are executed with both system gcc and clang in the %check section.

That's cool, I didn't know you could do that in an RPM spec file.  Great idea.

> I expect "you will start with a 0.5 or something". :)

Yeah, that's still my plan.  0.5 seems like a reasonable number to me given the state of things.

Also, do you guys want to keep the packaging in the SIMDe repo?  I'm not sure what the best practice is, but I'd be willing to do it if that's what you want.

**Comment by nemequ** [comment]
@junaruga I haven't packaged an RPM in years and I was never really an expert, so these might be completely insane, but I have a few questions about the spec file:

First, I see you're just installing to include/simde.  Maybe we should use include/simde-0.5/simde instead, and provide a simde-0.5.pc with `Cflags: -I${includedir}/simde-0.5`?  Or do you think that would beover-complicating things?  Any backwards-incompatible changes should be relatively minor, so maye it's not necessary, but I'd be surprised if there weren't a couple minor API breaks before 1.0…

Also, I'm not really sure what the implications of having the RPM provide Hedley like that are.  Would this cause a problem if another package provided Hedley in the same way?  Since hedley.h will be in the simde subdirectory, does SIMDe really provide it?

Hedley is kind of a special header in that it can be included multiple times without conflicts.  If a newer version is included it redefines all the macros to the new version.  If an older version is included it is ignored.  Basically, it should be completely safe to include even in public headers without worrying about versions.

**Comment by mr-c** [comment]
On Wed, Apr 1, 2020, 08:16 Evan Nemerson <notifications@github.com> wrote:

> Do you need a build system to install the headers?
>
No, but that would still be nice. Especially if it installs any other
documentation on using SIMDe that would be relevant offline.

That's really the only blocker left on my list at this point. I played with
> switching the immediate mode arguments to int_fast*_t, and hated it. It
> was fine when the allowed ranges matched up with the types, bust most of
> the time an int_fast8_t was still too big and it just ended up looking
> awkward.
>
> Other than that, I just want to get PPC64LE working on clang. I have a
> patch for µnit working for that, I just need to do some additional testing
> on the µnit side to make sure it doesn't break anything. I can probably be
> done with that tomorrow (unless it ends up horribly breaking something, but
> I think it's pretty safe).
>
Okay!

> You can see the simde's unit tests are executed with both system gcc and
> clang in the %check section.
>
> That's cool, I didn't know you could do that in an RPM spec file. Great
> idea.
>
Nice. I'll add this to the Debian package as a package build time test and
a post packaging "autopkgtest" that our CI systems will run whenever the
binary packages for the compilers are updated.

> I expect "you will start with a 0.5 or something". :)
>
> Yeah, that's still my plan. 0.5 seems like a reasonable number to me given
> the state of things.
>
👍

> Also, do you guys want to keep the packaging in the SIMDe repo? I'm not
> sure what the best practice is, but I'd be willing to do it if that's what
> you want.
>
Debian packaging is best managed on our side. People can still download the
source Debian package and customize as they wish.

https://salsa.debian.org/med-team/simde

**Comment by junaruga** [comment]
> That's cool, I didn't know you could do that in an RPM spec file. Great idea.

@nemequ Yeah, that enables us to test simde on 6 native CPU architectures (x86_64, aarch64, ppc64le, s390x, i686, armv7hl) in Fedora before releasing the RPM package.
As Debian has more CPU architectures such as riscv64, mips64el and sparc64 and etc, it's really great for @mr-c add the unit tests in the Debian. :)

> Also, do you guys want to keep the packaging in the SIMDe repo? I'm not sure what the best practice is, but I'd be willing to do it if that's what you want.

Thanks for considering this. Right now keeping the packaging file on the upstream repo is not the best practice in Fedora.
My [fedora-simde repo](https://github.com/junaruga/fedora-simde) currently keeping the packaging file is for a temporary use until we will put the file to the Fedora's official repo.
After putting the file to the official repo, I will delete the temporary repository.

The Fedora's official repo URL will be `https://src.fedoraproject.org/rpms/simde` like for example Fedora's zlib repo: [https://src.fedoraproject.org/rpms/zlib](https://src.fedoraproject.org/rpms/zlib) .

However there is a service [Packit-as-a-Service](https://packit.dev/packit-as-a-service/) that we can test a GitHub's repo accessing Fedora's build system at the pull-request timing via the packaging file (`simde.spec`) on the Github repo. I am not familiar with the service. But it's worth to investigate it, if you like it.

**Comment by junaruga** [comment]
> I haven't packaged an RPM in years and I was never really an expert, so these might be completely insane, but I have a few questions about the spec file:

@nemequ Sure, I am happy to answer it.

>  First, I see you're just installing to include/simde. Maybe we should use include/simde-0.5/simde instead, and provide a simde-0.5.pc with Cflags: -I${includedir}/simde-0.5? Or do you think that would beover-complicating things? Any backwards-incompatible changes should be relatively minor, so maye it's not necessary, but I'd be surprised if there weren't a couple minor API breaks before 1.0…

I do not think that the `include/simde-0.5/simde` is commonly used in Fedora.

Checking my local environment, here is the number of the directories `/usr/include/somethingN`.

```
$ find /usr/include -maxdepth 1 -type d -name "*[0-9]" | wc -l
16
```

Here is the number of the directories `/usr/include/something`.

```
$ find /usr/include -maxdepth 1 -type d | wc -l
98
```

Your concern about the backwards-incompatible changes is reasonable.
When we upgrade the upstream version in the RPM package, we care about other RPM packages that requires the RPM package. If it is needed, we communicate with the RPM maintainers in advance.

```
$ dnf repoquery --whatrequires zlib-devel --disablerepo=* --enablerepo=rawhide
```

And the package maintainer can detect a build error soon when a depending RPM package's version is changed, and the building and running unit tests are failed immediately automatically by a system. Here is [an example of `bowite` RPM package page](https://koschei.fedoraproject.org/package/bowtie). So, they will fix the build failure such as patching a file in the RPM spec file. And because of the reason, adding unit tests in `%check` section of the RPM spec file is encouraged.

> Also, I'm not really sure what the implications of having the RPM provide Hedley like that are. 

The syntax `Provides: bundled(hedley) = %{hedley_version}` is to indicate the bundling library. You can see [this part](https://docs.fedoraproject.org/en-US/packaging-guidelines/#bundling) in the Fedora packaging guidlines. Ideally it is encouraged to create another RPM `hedley` in this case, then `simde` use the `hedley` RPM in it. But I wanted to release `simde` RPM at first, I added it in `simde.spec`.

> Would this cause a problem if another package provided Hedley in the same way? Since hedley.h will be in the simde subdirectory, does SIMDe really provide it?

As far as I know, it's not problem. The syntax `Provides: bundled(<libname>) = <version>` is just a note. SIMDe does not really provide Hedley RPM by the syntax.

> Hedley is kind of a special header in that it can be included multiple times without conflicts. If a newer version is included it redefines all the macros to the new version. If an older version is included it is ignored. Basically, it should be completely safe to include even in public headers without worrying about versions.

Sure. It's good to know it.

**Comment by nemequ** [comment]
I just added Meson support, which includes installing headers to /usr/include/simde.  Only common and x86 headers are installed, not arm,

I also added `SIMDE_VERSION_MAJOR`/`SIMDE_VERSION_MINOR`/`SIMDE_VERSION_MICRO`/`SIMDE_VERSION` macros to simde-common.h.

I think I'm ready for a 0.5 release, unless anyone else has any other ideas?

I'm also thinking I may as well try to use it as an opportunity to generate some publicity for the project.  I'll try to put together a good release announcement, and send/submit it to a few places (in addition to the [mailing list](https://groups.google.com/forum/#!forum/simde) that nobody uses).  Ideas welcome; all I can really think of is Phoronix, and that could be a bit of a double-edged sword…

**Comment by mr-c** [comment]
I had no idea there was a mailing list! In fact, following that link gives me the message

> This group either doesn't exist, or you don't have permission to access it. If you're sure this group exists, contact the Owner of the group and ask them to give you access.

I think submitting to Hacker News (news.ycombinator..) could be useful as well.

**Comment by junaruga** [comment]
I also did not know the mailing list. And I got the message as @mr-c as well.

> I think submitting to Hacker News (news.ycombinator..) could be useful as well.

I think it's a great idea.

I think creating an issue or wiki page to share Linux distributions packaging status later could be useful, because some people might come here to look for the clues to create simde package for other Linux distributions such as Ubuntu, SuSE, Arch Linux, Gentoo Linux, and etc.

It's like this.

---

## Linux distribution packaging status

* [Debian](https://salsa.debian.org/med-team/simde): working in progress
* [Fedora](https://github.com/junaruga/fedora-simde): working in progress

Note: we do not support Mac environment. (?)

---

**Comment by nemequ** [comment]
> I had no idea there was a mailing list!

I just linked to it in the README then left it alone, I didn't really do anything to publicize it.

Similarly, I created a [gitter channel](https://gitter.im/simd-everywhere/community) yesterday that I haven't even added to the README yet (I'm still playing with it).  I should really find a more prominent way to feature them in the README.

> In fact, following that link gives me the message

Wow, oops.  The group was set so anyone can view topics, post, or join the group, but it was only visible to members of the group.  It should be working now; I tried viewing it from a tab where I wasn't logged in to Google and it now acknowledges that there is actually a list there.

> I think submitting to Hacker News (news.ycombinator..) could be useful as well.

Someone already [submitted](https://news.ycombinator.com/item?id=14177635) a link to the project a few years ago, I don't think I can submit it again.  Maybe a link to the release announcement, though.  /r/programming on Reddit, too (IIRC I submitted it to /r/C_programming but not /r/programming).

> I think creating an issue or wiki page to share Linux distributions packaging status later could be useful

Sure, that's probably a good idea.  I'll add something to the wiki later today.

We [do support](https://travis-ci.org/github/nemequ/simde/jobs/671139450) macOS.

**Comment by junaruga** [comment]
> Someone already submitted a link to the project a few years ago, I don't think I can submit it again. Maybe a link to the release announcement, though. /r/programming on Reddit, too (IIRC I submitted it to /r/C_programming but not /r/programming).

Okay. I agree.

> Sure, that's probably a good idea. I'll add something to the wiki later today.

Thanks!

> We do support macOS.

Ah okay. I forgot the case was in Travis.

Good news is now simde is in review status in Fedora. I opened the ticket yesterday :)
https://bugzilla.redhat.com/show_bug.cgi?id=1823001

**Comment by mr-c** [comment]
A snapshot package of SIMDe has been accepted into Debian Unstable "sid"!

https://tracker.debian.org/news/1117911/accepted-simde-000git20200402-1-source-amd64-into-unstable-unstable/

I'll freshen the snapshot in the next day or two

**Comment by junaruga** [comment]
It's great to hear that! I also passed the review today in Fedora. So we can ship simde to both Debian and Fedora at the same time. :)

**Comment by junaruga** [comment]
Just note I am going to ship the Fedora SIMDe package with version `0.0.0` aligning the Debian package version.

**Comment by junaruga** [comment]
A package of SIMDe is available on Fedora unstable (rawhide = f33) now!
I am trying to build it into f32, f31, f30, epel8 (= rhel8) and epel7 (= rhel7).
https://src.fedoraproject.org/rpms/simde

**Comment by nemequ** [comment]
Sorry, I forgot to close this, but 0.5.0 has been out for a while now.  I'll probably do another one soon (I think 0.6.0 would be warranted given the huge improvements in NEON support, plus SVML and AVX2, and lots of other performance improvements).
